File: romantodecimal.py
MAX_ADDITIVE = 3
MAX_SUBTRACTIVE = 1

ROMAN = ["M", "D", "C", "L", "X", "V", "I"]
DECIMAL = [1000, 500, 100, 50, 10, 5, 1]

def R2D(letter):
    """
    Converts Roman character to decimal integer
    """
    if letter == None:
        return 0
    else:
        return DECIMAL[ROMAN.index(letter)]

def isEven(num):
    """
    Returns True if even, False if odd
    """
    if num % 2 == 0:
        return True
    else:
        return False

def getNumDigits(letter):
    """
    Returns number of digits in integer
    """
    return len(str(R2D(letter)))

def getDigitsDifference(A, B):
    """
    Returns difference between number of digits for A and B\n
    eg A=10(2), B=1000(4), output=2   
    """
    da, db = getNumDigits(A), getNumDigits(B)
    return db-da

def getFirstDigit(letter):
    """
    Returns first integer digit of input Roman numeral
    """
    num = R2D(letter)
    while num >= 10:
        num /= 10
    return num

def isSameDigit(letterA,letterB):
    """
    Returns True if the first digits match, False otherwise
    """
    return getFirstDigit(letterA) == getFirstDigit(letterB)

def findOperation(A, B=None, C=None): # UP DOWN
    """
    A       A>B=C   A>B>C   A>C>B
    A+B     A=B>C   
    B-A     B>A=C   B>A>C
    A+B+C   A=B=C
    """
    da, db, dc = R2D(A), R2D(B), R2D(C)
    dac, dab, dbc = getDigitsDifference(A,C), getDigitsDifference(A,B), getDigitsDifference(B,C)

    if da < dc:
        raise Exception("Error: Incorrect order of ascension (C) - {} {} [{}]".format(A,B,C))
    
    if da == dc:
        if getFirstDigit(A) == 5:
            raise Exception("Error: Two 5s of same kind - [{}] {} [{}]".format(A,B,C))
        if db > da:
            raise Exception("Error: Incorrect ascension (B) - {} [{}] {}".format(A,B,C))
        if dab < -1:
            raise Exception("Error: Incorrect descension (B) - {} [{}] {}".format(A,B,C))
    
    if db > da:
        if da == dc:
            if getFirstDigit(A) == 5:
                raise Exception("Error: Two 5s of same kind - [{}] {} [{}]".format(A,B,C))
            if dab < -1:
                raise Exception("Error: Incorrect descension (B) - {} [{}] {}".format(A,B,C))        
        if getDigitsDifference(A,B) >= 1:
            if getDigitsDifference(A,B) > 1 or (getDigitsDifference(A,B) == 1 and not isSameDigit(A, B)):
                raise Exception("Error: Incorrect ascension (B) - {} [{}] {}".format(A,B,C))
        return 1

    if dc > db:
        if getFirstDigit(B) != 1:   # Only 1s can come before higher digits
            raise Exception("Error: Wrong digit for subtraction - {}".format(B))
        if getDigitsDifference(B,C) > 0:
            if getDigitsDifference(B,C) > 1:
                raise Exception("Error: Incorrect order of ascension (B^C) - {} {}".format(A,B))
            elif not isSameDigit(B,C):
                raise Exception("Error: Wrong subtraction combination - {} {}".format(A,B))   
    if da >= db:
        return 0

def countLetters(input, letter):
    """
    Returns number of occurrences of input letter within input string
    """
    n = 0
    for d in input:
        if letter == d:
            n += 1
    return n

def validateAllLetters(input):
    """
    Checks if all letters in input string are valid Roman numerals
    """
    for letter in input:
        if letter not in ROMAN:
            raise Exception("Letter {} not valid".format(letter))

def validateRuns(input):
    """
    Checks that input string does not contain more than a set amount of the same character in succession
    """
    run = 1
    prev_letter = ""
    for letter in input:       
        if letter == prev_letter:
            run += 1
        else:
            run = 1
                
        if run > MAX_ADDITIVE:
            raise Exception("Exceeds max run length for {}".format(letter))
        
        prev_letter = letter
        
def processInput(user_input):
    """
    Takes string input and converts it to integer number output
    """

    output = 0

    validateAllLetters(user_input)
    validateRuns(user_input)

    while len(user_input) > 0:

        a = user_input[0]
        b = None
        c = None

        if len(user_input) >= 2:
            b = user_input[1]
        if len(user_input) >= 3:
            c = user_input[2]

        n = findOperation(a, b, c)

        if n == 0:
            output += R2D(a)
            user_input = user_input[1:]

        elif n == 1:
            output -= R2D(a)
            output += R2D(b)
            user_input = user_input[2:]

    return output


#############################################################

user_input = ""

while user_input == "":
    user_input = input("Enter numeral or Q: ")
    if user_input != "Q":
        try:
            print("{} = {}".format(user_input, processInput(user_input)))
        except Exception as err:
            print(err.args)
    user_input = ""

File: romantodecimal.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

const int MAX_ADDITIVE = 3;
const int MAX_SUBTRACTIVE = 1;

string roman = "MDCLXVI";
int decimal [7] = {1000, 500, 100, 50, 10, 5, 1};

int findLetterIndex(char letter){
    int size = roman.size();
    for(int i = 0; i < size; i++){
        if(roman[i] == letter){
            return i;
        }
    }
    return -1;
}

int getFirstDigit(char letter){
    int n = decimal[findLetterIndex(letter)];
    while(n >= 10){
        n /= 10;
    }
    return n;
}

bool isSameDigit(char a, char b){
    int one = getFirstDigit(a);
    int two = getFirstDigit(b);
    if(one==two){
        return true;
    } else {
        return false;
    }
}

int getNumDigits(char letter){
    int index = findLetterIndex(letter);
    if(index != -1){
        int num = decimal[index];
        int n = 1;
        while(num >= 10){
            n += 1;
            num /= 10;
        }
        return n;
    } else {
        throw runtime_error("letter not found");
    }
}

int getNumDigitsDifference(char A, char B){
    int numA = getNumDigits(A);
    int numB = getNumDigits(B);
    return numB-numA;
}

int findOperation(char a, char b, char c){
    int da = decimal[findLetterIndex(a)];
    int db = decimal[findLetterIndex(b)];
    int dc = decimal[findLetterIndex(c)];

    if(da<dc){
        throw runtime_error("invalid order of ascension");
    }

    if(da==dc){
        if(getFirstDigit(a) == 5){
            throw runtime_error("two 5s of same kind");
        }
        if(db > da){
            throw runtime_error("incorrect ascension");
        }
        if(getNumDigitsDifference(a,b) > 1){
            throw runtime_error("incorrect descension");
        }
    }

    if(db>da){
        if(da==dc){
            if(getFirstDigit(a) == 5){
                throw runtime_error("two 5s of same kind");
            }
            if(getNumDigitsDifference(a,b) < -1){
                throw runtime_error("incorrect descension");
            }
        }
        if(getNumDigitsDifference(a,b) > 1){
            throw runtime_error("incorrect descension");
        }
        if(getNumDigitsDifference(a,b) == 1 && !isSameDigit(a,b)){
            throw runtime_error("incorrect descension");
        }
        return 1;
    }

    if(dc>db){
        if(getFirstDigit(b) != 1){
            throw runtime_error("wrong digit for subtraction");
        }
        if(getNumDigitsDifference(b,c) > 0){
            if(getNumDigitsDifference(b,c) > 1 || !isSameDigit(b,c)){
                throw runtime_error("incorrect order of ascension");
            }
        }
    }

    if(da>=db){
        if(da==db){
            if(getFirstDigit(a) == 5){
                throw runtime_error("two 5s of same kind");
            }
        }
        return 0;
    }
    return -1;
}

void validateLetters(string in){
    for(int i = 0; i < in.size(); i++){
        if(roman.find(in[i]) == string::npos){
            throw runtime_error("invalid character");
        }
    }
}

void validateRuns(string in){
    int run = 1;
    char prev_letter = ' ';
    for(int i = 0; i < in.size(); i++){
        char letter = in[i];
        if(letter == prev_letter){
            run += 1;
        } else {
            run = 1;
        }

        if(run > MAX_ADDITIVE){
            throw runtime_error("exceeds max run length");
        }

        prev_letter = letter;
    }
}

int calculate(string in){

    validateLetters(in);
    validateRuns(in);

    int output = 0;

    while(in.size() > 0){
        char a = in[0];
        char b = 0, c = 0;

        if(in.size() >= 2){
            b = in[1];
        }
        if(in.size() >= 3){
            c = in[2];
        }

        int n = findOperation(a,b,c);

        if(n==0){ // Addition
            output += decimal[findLetterIndex(a)];
            in = in.substr(1);
        } else if(n==1){ // Subtraction
            output -= decimal[findLetterIndex(a)];
            output += decimal[findLetterIndex(b)];
            in = in.substr(2);
        }

    }
    return output;
}

int main(){
    string user_input = "none";

    while(user_input == "none"){
        cout << "Enter numeral or Q: ";
        cin >> user_input;

        if(user_input != "Q"){
            try{
                int final_output = calculate(user_input);
                cout << user_input << " = " << final_output << endl;
            } catch(const runtime_error& err){
                cout << err.what() << endl;
            }
            user_input = "none";
        }
    }

    return 0;
}